include/vecmath/distance.h

```diff
@@ -262,9 +267,24 @@ namespace vm {
         const auto D = a * c - b * b;
 
         if (is_zero(D, constants<T>::almost_zero())) {
-            const auto f = dot(w, v);
-            const auto z = w - f * v;
-            return line_distance<T>::Parallel(squared_length(z));
+            // parallel case
+            const T p1OnR = distance_to_projected_point(r, p1);
+            const T p2OnR = distance_to_projected_point(r, p2);
+            const T perpendicularDistSquared = squared_distance(point_at_distance(r, p1OnR), p1);
+
+            if (p1OnR < static_cast<T>(0) && p2OnR < static_cast<T>(0)) {
+                // segment completely behind ray
+                const T segmentLength = sqrt(a);
+                const auto& pointOnSegmentClosestToRay = (p1OnR > p2OnR) ? p1 : p2;
+                return line_distance<T>::Parallel(0, squared_distance(r.origin, pointOnSegmentClosestToRay), segmentLength);
+            } else if (p1OnR > static_cast<T>(0) && p2OnR > static_cast<T>(0)) {
+                // segment completely in front of ray origin
+                return line_distance<T>::Parallel(min(p1OnR, p2OnR), perpendicularDistSquared, 0);
```

Comment:
Likewise here, the third value must be either 0 or `segmentLength`.

Reply:
Right